### ISSUE TITLE
[Enhancement] Some enhancements for transformer encoder

### ISSUE THREAD

**Issue by 842974287** [NVIDIA/FasterTransformer#98]
Hi,

We had some enhancements over the base FasterTransformer to accommodate our use cases. It would be great if we can upstream those enhancements so that we don't have to migrate those changes for future releases. I saw in the contributing rules, enhancements should be reviewed by issue request prior to code reviews. So I listed the enhancements below and would appreciate if you guys can take a look.

1. Add support for encoders that have head_dim (head_num * size_per_head) larger than 1024. Currently `add_bias_input_layernorm` and `add_QKV_bias_rebuild_padding` kernels used in bert encoder don't support head_dim larger than 1024. The changes extend these kernels to support that.
2. Add support for different mlp hidden dimension size. Currently in bert encoder feed forward part, the hidden dim of the first fc is hardcoded to 4 * head_dim * head_size. We added another field to pass in the size of hidden dimension.
3. Add support for customizing scalar in softmax kernel. Some of our models have large projection weights for query. This leads to overflow in fp16. To prevent that we scale the q_proj weights in the models and reduce the scalar that is used in forward.
4. Add another code path for normalizing before attention. In the tensor2tensor code they suggest that learning is more robust when preprocessing each layer with layernorm and postprocessing with `dropout -> add residual`. In fairseq transformer encoder layer module, they have an option to enable this path. The changes also added corresponding path to FasterTransformer bert encoder.

I have the code ready in [my branches](https://fburl.com/fr36uz38). Let me know if those changes make sense and I'll submit PRs for code review. Thanks!

**Comment by byshiue** [comment]
It's a great work. But I cannot open the link before. From your repo of FasterTransformer, there are many possible branch, which one should I check? 
Besides, for the forth enhancement above, we have provided another structure of encoder in `standard_encoder.h`, which follow the model strcture of Attention is All You Need, but not Bert. Is is same to this enhancement?

**Comment by 842974287** [comment]
Hi @byshiue , thanks for taking a look! I listed the commit for each enhancements below.

1. https://github.com/842974287/FasterTransformer/commit/dacb3ceed52d6cdb59f10adc6fa02f615da9084a
2. https://github.com/842974287/FasterTransformer/commit/0e83cfe9cb0ab266d47db44942188434f15dcd63
3. https://github.com/842974287/FasterTransformer/commit/75ec8b04bcbfa9201adb181d9997cbb3fb85682a

Emitting the 4th as it's indeed the same as `standard_encoder.h`. Thanks for pointing that out! I guess then do you think it would make sense to add an option in FTencoder in `encoder.h` to control which encoder we use in the forward (switching between stardard_encoder and bert)?

**Comment by byshiue** [comment]
I have give some advices for these commits. After you fix them, you can use https://github.com/842974287/FasterTransformer/tree/main/sample/tensorflow/unit_test#encoder to check the correctness, and prepare submit pull requests.
For 4th, we will consider a better way to switch.

**Comment by byshiue** [comment]
Those features are supported in latest release.
